goss-testing/scripts/data-validator.py

```diff
@@ -0,0 +1,289 @@
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+import re
+import subprocess
+import sys
+import ipaddress
+import socket
+import json
+import os
+
+
+def print_err(*a):
+  print(*a, file = sys.stderr)
+
+
+def get_data():
+  """Get data from data.json or BSS.
+  """
+  hostname = socket.gethostname()
+
+  if os.path.isfile("/etc/pit-release"):
+    try:
+      print_err(f"\nRunning on pit node: {hostname}. Querying data.json...")
+      print_err("------------------------------------------------------------")
+      with open("/var/www/ephemeral/configs/data.json", 'r') as file:
+        return json.load(file)
+    except Exception as e:
+      print_err(str(e))
+      return 1
+  else:
+    try:
+      print_err(f"\nRunning on node: {hostname}. Querying BSS...")
+      print_err("------------------------------------------------------------")
+      command = [ "cray", "bss", "bootparameters", "list", "--format", "json" ]
+      bss_proc = subprocess.Popen(command, stdout=subprocess.PIPE)
+      return json.loads(bss_proc.stdout.read())
+    except Exception as e:
+      print_err(str(e))
+      return 1
+
+
+def user_data(data):
+  """Returns the 'user-data' blobs from data.json or BSS.
+  """
+  filtered_data = []
+
+  # if we're using BSS data
+  if isinstance(data, list):
+    for blob in data:
+        if isinstance(blob['cloud-init']['user-data'], dict):
+          filtered_data.append(blob['cloud-init']['user-data'])
+
+  # if we're using basecamp data
+  elif isinstance(data, dict) and "Global" in data:
+    for blob in data:
+      if blob != "Global":
+        filtered_data.append(data[blob]['user-data'])
+  
+  return filtered_data
+
+
+def drill_down_data(data, desired_keys):
+  for i in range(0, ( len(desired_keys) + 1), 1):
+    if i == len(desired_keys):
+      return saved_data
+    if i == 0:
+      saved_data = data[desired_keys[i]]
+    elif i != 0:
+      saved_data = saved_data[desired_keys[i]]
+
+
+def are_valid_ip_masks(data, desired_keys):
```

Comment:
This function passes if the data (or the filtered_data) is empty. Is this intended? Note -- I don't know the answer here. If it's okay for the data to be missing, and we're just wanting to make sure it's valid if it is present, then the current behavior is okay.

Reply:
I'm including a check for empty data and a check for confirming that desired_keys exist.